### DISCUSSION TITLE
Request support for SQLAlchemy-Redshift

### DISCUSSION THREAD

**Discussion by nlm4145** [sqlalchemy/sqlalchemy#11950] [Ideas]
### Describe the use case

The maintainers of [sqlalchemy-redshift](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift) have left developers high and dry when it came to the release of SQLAlchemy 2.0; no communication on when updates would be made to the library to support SA 2 +, no merging of PRs, et cetera ([example git issue](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/issues/264)).

Some fixes such as `engine.dialect._set_backslash_escapes = lambda _: None` have been implemented that allow basic functionality of SA but other critical features such as MetaData reflection still do not work.

Basically it seems that the dialect for postgres has changed so much since 1.4 that it is now completely incompatible with redshift. Is there anything the maintainers of sqlalchemy can do, or suggest, to re-integrate the redshift dialect to work w/ SA 2+?

I have shared examples of issues in the "Example Use" section.

### Databases / Backends / Drivers targeted

Redshift / PostgreSQL

### Example Use

```
from sqlalchemy import create_engine, MetaData

engine = create_engine(f'redshift+redshift_connector://user:password@localhost:5439/mydb', future=True, connect_args={'sslmode': 'verify-ca'})
if hasattr(engine.dialect, "_set_backslash_escapes"):
    engine.dialect._set_backslash_escapes = lambda _: None

meta = MetaData()
meta.reflect(bind=engine, schema="mydb")
```

returns this error:
```
sqlalchemy.exc.ProgrammingError: (redshift_connector.error.ProgrammingError) {'S': 'ERROR', 'C': '42P01', 'M': 'relation "pg_catalog.pg_collation" does not exist', 'F': '/home/ec2-user/padb/src/pg/src/backend/catalog/namespace.c', 'L': '240', 'R': 'LocalRangeVarGetRelid'}
[SQL: SELECT pg_catalog.pg_type.typname AS name, pg_catalog.format_type(pg_catalog.pg_type.typbasetype, pg_catalog.pg_type.typtypmod) AS attype, NOT pg_catalog.pg_type.typnotnull AS nullable, pg_catalog.pg_type.typdefault AS "default", pg_catalog.pg_type_is_visible(pg_catalog.pg_type."oid") AS visible, pg_catalog.pg_namespace.nspname AS schema, domain_constraints.condefs, domain_constraints.connames, pg_catalog.pg_collation.collname 
FROM pg_catalog.pg_type JOIN pg_catalog.pg_namespace ON pg_catalog.pg_namespace."oid" = pg_catalog.pg_type.typnamespace LEFT OUTER JOIN pg_catalog.pg_collation ON pg_catalog.pg_type.typcollation = pg_catalog.pg_collation."oid" LEFT OUTER JOIN (SELECT pg_catalog.pg_constraint.contypid AS contypid, array_agg(pg_catalog.pg_get_constraintdef(pg_catalog.pg_constraint."oid", %s)) AS condefs, array_agg(CAST(pg_catalog.pg_constraint.conname AS TEXT)) AS connames 
FROM pg_catalog.pg_constraint 
WHERE pg_catalog.pg_constraint.contypid != %s GROUP BY pg_catalog.pg_constraint.contypid) AS domain_constraints ON pg_catalog.pg_type."oid" = domain_constraints.contypid 
WHERE pg_catalog.pg_type.typtype = %s ORDER BY pg_catalog.pg_namespace.nspname, pg_catalog.pg_type.typname]
[parameters: (True, 0, 'd')]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

We get variations of this error whether we use redshift_connector or psycopg.

### Additional context
In absence of an actual redshift cluster for development, I imagine one could easily use localstack.

_No response_

**Comment by CaselIT** [comment]
Hi,

> Basically it seems that the dialect for postgres has changed so much since 1.4 that it is now completely incompatible with redshift. Is there anything the maintainers of sqlalchemy can do, or suggest, to re-integrate the redshift dialect to work w/ SA 2+?

As far as I know there is no free version of redshift, so that complicates quite a bit any initiative from sqlalchemy maintainers. It's not possible to develop and maintain a dialect without access to the db to run ci pipelines.

> In absence of an actual redshift cluster for development, I imagine one could easily use localstack.

Seems to be a commercial offering too, so see point above.

In any case the sqlalchemy team does not currently seek to maintain other dialects at this time.

Regarding changes between 1.4 and v2, I'm sure it's possible to take the necessary bit from the pg dialect in the 1.4 branch and use them for the redshift dialect.

Moving to discussion

**Comment by zzzeek** [comment]
it's not unusual for an open source project to be abandoned.   there's a standard procedure for this which starts with selecting someone who wants to pick up the work, then either getting access from the project's authors, or forking it.     for sqlalchemy-redshift I would imagine there might be some involvement from AWS employees directly and it would be poor form of them to not even reply to requests to transfer the project to new owners
